File: backend/GulhanKahvalti.API/Services/Implementations/AuthService.cs
using GulhanKahvalti.API.DTOs.Auth;
using GulhanKahvalti.API.Helpers;
using GulhanKahvalti.API.Models;
using GulhanKahvalti.API.Repositories.Interfaces;
using GulhanKahvalti.API.Services.Interfaces;

namespace GulhanKahvalti.API.Services.Implementations;

public class AuthService(
    IUserRepository userRepository,
    JwtTokenGenerator jwtTokenGenerator) : IAuthService
{
    public async Task<AuthResponseDto> RegisterAsync(RegisterRequestDto request)
    {
        var normalizedEmail = request.Email.Trim().ToLowerInvariant();

        if (await userRepository.ExistsByEmailAsync(normalizedEmail))
        {
            throw new InvalidOperationException("Email is already registered.");
        }

        var user = new User
        {
            FullName = request.FullName.Trim(),
            Email = normalizedEmail,
            PasswordHash = PasswordHasher.HashPassword(request.Password),
            Role = "User",
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        await userRepository.CreateAsync(user);

        return CreateAuthResponse(user);
    }

    public async Task<AuthResponseDto> LoginAsync(LoginRequestDto request)
    {
        var normalizedEmail = request.Email.Trim().ToLowerInvariant();
        var user = await userRepository.GetByEmailAsync(normalizedEmail);

        if (user is null || !PasswordHasher.VerifyPassword(request.Password, user.PasswordHash))
        {
            throw new UnauthorizedAccessException("Invalid email or password.");
        }

        return CreateAuthResponse(user);
    }

    private AuthResponseDto CreateAuthResponse(User user)
    {
        return new AuthResponseDto
        {
            UserId = user.Id,
            FullName = user.FullName,
            Email = user.Email,
            Role = user.Role,
            Token = jwtTokenGenerator.GenerateToken(user)
        };
    }
}

File: frontend/gulhan-kahvalti-client/src/services/authService.ts
import { api } from './api'
import type { AuthResponse, LoginRequest, RegisterRequest, User } from '../types'

function mapAuthResponseToUser(response: AuthResponse): User {
  return {
    id: response.userId,
    fullName: response.fullName,
    email: response.email,
    role: response.role,
  }
}

export const authService = {
  async login(request: LoginRequest) {
    const { data } = await api.post<AuthResponse>('/auth/login', request)
    return { auth: data, user: mapAuthResponseToUser(data) }
  },

  async register(request: RegisterRequest) {
    const { data } = await api.post<AuthResponse>('/auth/register', request)
    return { auth: data, user: mapAuthResponseToUser(data) }
  },

  async me() {
    const { data } = await api.get<User>('/auth/me')
    return data
  },
}
